### PR TITLE
Add gRPC authentication using google.auth

### PR DESCRIPTION
**Pull request by jonparrott**
@bjwatson @geigerj PTAL.

This just adds the module, it's not yet wired up in `google.gax.grpc`. That'll be the next PR.

### CODE REVIEW

**Comment by jonparrott** [comment]
@geigerj I'm not sure how to fix 'pylint-no-test-deps' or even why that's a thing.

File: google/gax/_grpc_google_auth.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+from __future__ import absolute_import
+
+import google.auth
+import google.auth.credentials
+import google.auth.transport.grpc
+
+try:
+    import google.auth.transport.requests
+    # pylint: disable=invalid-name
+    # PyLint recognizes this as a class, but we're aliasing it as a factory
```
Comment (dhermes) [inline google/gax/_grpc_google_auth.py]: Tis Pylint good sir

**Comment by geigerj** [comment]
@jonparrott no-test-deps ensures that you update `requirements.txt` or `setup.py` (I forget which) and not just `test-requirements.txt` when you add a new dependency.

**Comment by jonparrott** [comment]
> no-test-deps checks that you update requirements.txt or setup.py (I forget which) and not just test-requirements.txt when you add a new dependency.

These are optional dependencies. Good show. :)

**Comment by jonparrott** [comment]
@geigerj that is to say - these dependencies are optional so `pylint-no-test-deps` will always fails. There shouldn't be any duplication between `setup.py` and `test-requirements.txt`, so can we just remove `pylint-no-test-deps`?

**Comment by geigerj** [comment]
Hm... that makes our checks less strict, because if you add a non-optional dependency to `test-requirements.txt` instead of to `setup.py`, the Travis build will pass but you'll get a runtime import error if you install and run the package. Instead, what if we explicitly `--ignore` this file in the pylint command for `test-requirements.txt`?

**Comment by jonparrott** [comment]
@geigerj this is actually the first time I've ever seen a project do this.

**Comment by geigerj** [comment]
@bjwatson Any thoughts on whether we should kill this test environment? We created it in response to #108 originally.

**Comment by geigerj** [comment]
I chatted to @bjwatson offline, and we'd like to keep `no-test-deps`. Just add this file to the `--ignore` list here: https://github.com/googleapis/gax-python/blob/master/tox.ini#L34.

**Comment by geigerj** [comment]
Also, should the optional dependency be added to `setup.py` [like this](https://github.com/googleapis/api-client-staging/blob/8e8e7a40b961a10f94b377dc0e467221633b7a09/generated/python/googleapis-common-protos/setup.py#L17)?

**Comment by jonparrott** [comment]
> Just add this file to the --ignore list here: 

will do (in a hackathon today, so will follow-up on monday)

> Also, should the optional dependency be added to setup.py like this?

Not yet.

**Comment by jonparrott** [comment]
pylint ignores added, PTAL.

File: tox.ini [hunk 2]
```diff
@@ -23,20 +23,40 @@ commands = flake8 --max-complexity=10 google test --ignore=E501 --exclude=src-ge
 [testenv:pylint-errors]
 deps = pylint
        -r{toxinidir}/test-requirements.txt
-commands = pylint -f colorized -E google test --ignore=src-gen
+commands = pylint \
+  -f colorized \
+  -E \
+  google test \
+  --ignore=src-gen,_grpc_google_auth.py,test__grpc_google_auth.py
```
Comment (geigerj) [inline tox.ini]: I thought we only needed this change in `pylint-no-test-deps`; is there a reason to ignore it in the environments that do have a dependency on `test-requirements.txt`?

Comment (jonparrott) [inline tox.ini]: Ah, right. Fixed. (having this many different pylint runs is a little unusual)

**Comment by jonparrott** [comment]
I don't have access to do so, if you want to add me as a contributor on this repo then I can.

**Comment by geigerj** [comment]
Ah, I think omaray@ would need to do that.

**Comment by jonparrott** [comment]
Thanks! Hopefully just one more PR to finish this up. 👍

**Pull request merged**